README.md
# color_pick

A new Flutter application.

## Getting Started

This project is a starting point for a Flutter application.

A few resources to get you started if this is your first Flutter project:

- [Lab: Write your first Flutter app](https://flutter.dev/docs/get-started/codelab)
- [Cookbook: Useful Flutter samples](https://flutter.dev/docs/cookbook)

For help getting started with Flutter, view our
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.

![image](https://github.com/xfhomedream/color_pick/blob/master/1564236994665_450x900.gif)


CHANGELOG.md
[0.0.1] - TODO: Add release date.
Initial release.
[0.0.2]
Update example
[0.0.3]
Update example
[0.0.4]
Update example
[0.0.5]
Update example
[0.0.6]
Update example
[0.0.9]
Optimization code
[0.1.0]
support Null safety

test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:color_pick_view/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/main.dart
import 'package:color_pick_view/ring_page.dart';
import 'package:flutter/material.dart';

import 'circle_page.dart';
import 'color_pick.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title!),
      ),
      body: Column(
        children: <Widget>[
          ListTile(
            title: const Text("色盘选择器"),
            trailing: const Icon(Icons.arrow_forward_ios),
            onTap: () {
              Navigator.of(context).push(MaterialPageRoute(
                  builder: (BuildContext context) => new CirclePage()));
            },
          ),
          const Divider(),
          ListTile(
            title: const Text("色环选择器"),
            trailing: const Icon(Icons.arrow_forward_ios),
            onTap: () {  Navigator.of(context).push(MaterialPageRoute(
                builder: (BuildContext context) =>  RingPage()));},
          ),
        ],
      ),
    );
  }
}


lib/ring_page.dart
import 'package:flutter/material.dart';

import 'ring_color_pick.dart';

class RingPage extends StatefulWidget {
  const RingPage({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    // TODO: implement createState
    return RingPageState();
  }
}

class RingPageState extends State<RingPage> {
  Color currentColor = const Color(0xff0000ff);

  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: const Text("Color wheel selector"),
      ),
      body: Center(
          child: Column(
        children: <Widget>[
          ColorRingPickView(
            selectColor: const Color(0xff00eaff),
            selectRadius: 40,
            ringWidth: 30,
            selectRingColor: Colors.white,
            selectColorCallBack: (color) {
              setState(() {
                currentColor = color;
              });
            },
          ),
          Container(
            color: currentColor,
            height: 50,
            width: 50,
            child: const SizedBox(),
          )
        ],
      )),
    );
  }
}


lib/color_pick.dart
library color_pick;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'dart:math';

typedef SelectColor = void Function(Color color);

class ColorPickView extends StatefulWidget {
  final Size? size;
  final double? selectRadius;
  final double? padding;
  final Color? selectColor;
  final Color? selectRingColor;
  final SelectColor? selectColorCallBack;

  ColorPickView(
      {Key? key,
      this.size,
      this.selectColorCallBack,
      this.selectRadius,
      this.padding,
      this.selectRingColor,
      this.selectColor})
      : super(key: key) {
    assert(size == null || (size != null), '控件宽度太宽');
    assert(size == null || (size != null && size!.height == size!.width),
        '控件宽高必须相等');
  }

  @override
  State<StatefulWidget> createState() {
    return ColorPickState();
  }
}

class ColorPickState extends State<ColorPickView> {
  double? radius;
  double? selectRadius, padding;
  Color currentColor = const Color(0xff00ff);
  Color? selectRingColor;

  Offset? currentOffset;
  Offset? topLeftPosition;
  Offset? selectPosition;
  Size? screenSize, size;
  GlobalKey globalKey = new GlobalKey();
  bool isTap = false;

  @override
  Widget build(BuildContext context) {
    screenSize ??= MediaQuery.of(context).size;
    size = (widget.size ?? screenSize);
    selectRadius = (widget.selectRadius ?? 10);
    padding = (widget.padding ?? 40);
    selectRingColor = (widget.selectRingColor ?? Colors.black);
    assert((size != null && screenSize!.width >= size!.width), '控件宽度太宽');
    radius = size!.width / 2 - padding!;
    currentOffset ??= Offset(radius!, radius!);
    if (widget.selectColor != null && selectPosition == null) {
      _setColor(widget.selectColor!);
    }
    _initLeftTop();
    return GestureDetector(
      key: globalKey,
      child: SizedBox(
        width: size!.width,
        height: size!.width,
        child: Stack(
          alignment: Alignment.center,
          children: <Widget>[
            CustomPaint(
              painter: ColorPick(radius: radius),
              size: size!,
            ),
            Positioned(
              left: isTap
                  ? currentOffset!.dx -
                      (topLeftPosition == null
                          ? 0
                          : (topLeftPosition!.dx + selectRadius! / 2))
                  : (selectPosition == null
                      ? radius
                      : selectPosition!.dx + selectRadius! / 2),
              top: isTap
                  ? currentOffset!.dy -
                      (topLeftPosition == null
                          ? 0
                          : (topLeftPosition!.dy + selectRadius! / 2))
                  : (selectPosition == null
                      ? radius
                      : selectPosition!.dy + selectRadius! / 2),
              //这里减去80，是因为上下边距各40 所以需要减去还有半径
              child: Container(
                width: selectRadius,
                height: selectRadius,
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(selectRadius!),
                  border: Border.fromBorderSide(
                      BorderSide(color: selectRingColor!)),
                ),
                child: ClipOval(
                  child: Container(
                    color: currentColor,
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
      onTapDown: (e) {
        setState(() {
          isTap = true;
          _initLeftTop();
          if (!isOutSide(e.globalPosition.dx, e.globalPosition.dy)) {
            currentColor =
                getColorAtPoint(e.globalPosition.dx, e.globalPosition.dy);
            currentOffset = e.globalPosition;
            if (widget.selectColorCallBack != null) {
              widget.selectColorCallBack!(currentColor);
            }
          }
        });
      },
      onPanUpdate: (e) {
        isTap = true;
        _initLeftTop();
        setState(() {
          if (!isOutSide(e.globalPosition.dx, e.globalPosition.dy)) {
            currentOffset = e.globalPosition;
            currentColor =
                getColorAtPoint(e.globalPosition.dx, e.globalPosition.dy);
            if (widget.selectColorCallBack != null) {
              widget.selectColorCallBack!(currentColor);
            }
          }
        });
      },
    );
  }

  void _initLeftTop() {
    if (globalKey.currentContext != null) {
      final RenderBox? box =
          globalKey.currentContext!.findRenderObject() as RenderBox?;
      topLeftPosition = box?.localToGlobal(Offset.zero);
    }
  }

  bool isOutSide(double eventX, double eventY) {
    double x = eventX - (topLeftPosition!.dx + radius! + padding!);
    double y = eventY - (topLeftPosition!.dy + radius! + padding!);
    double r = sqrt(x * x + y * y);
    if (r >= radius!) return true;
    return false;
  }

  void _setColor(Color color) {
    //设置颜色值
    var hsvColor = HSVColor.fromColor(color);
    double r = hsvColor.saturation * radius!;
    double radian = hsvColor.hue / -180.0 * pi;
    _updateSelector(r * cos(radian), -r * sin(radian));
    currentColor = color;
  }

  void _updateSelector(double eventX, double eventY) {
    //更新选中颜色值
    double r = sqrt(eventX * eventX + eventY * eventY);
    double x = eventX, y = eventY;
    if (r > radius!) {
      x *= radius! / r;
      y *= radius! / r;
    }
    selectPosition = Offset(x + radius! + padding!, y + radius! + padding!);
  }

  Color getColorAtPoint(double eventX, double eventY) {
    //获取坐标在色盘中的颜色值
    double x = eventX - (topLeftPosition!.dx + radius! + padding!);
    double y = eventY - (topLeftPosition!.dy + radius! + padding!);
    double r = sqrt(x * x + y * y);
    List<double> hsv = [0.0, 0.0, 1.0];
    hsv[0] = (atan2(-y, -x) / pi * 180).toDouble() + 180;
    hsv[1] = max(0, min(1, (r / radius!)));
    return HSVColor.fromAHSV(1.0, hsv[0], hsv[1], hsv[2]).toColor();
  }
}

class ColorPick extends CustomPainter {
  Paint? mPaint;
  Paint? saturationPaint;
  final List<Color> mCircleColors = [];
  final List<Color> mStatColors = [];
  SweepGradient? hueShader;
  final double? radius;
  RadialGradient? saturationShader;

  ColorPick({this.radius = 10}) {
    mPaint = Paint();
    saturationPaint = Paint();
    mCircleColors.add(const Color.fromARGB(255, 255, 0, 0));
    mCircleColors.add(const Color.fromARGB(255, 255, 255, 0));
    mCircleColors.add(const Color.fromARGB(255, 0, 255, 0));
    mCircleColors.add(const Color.fromARGB(255, 0, 255, 255));
    mCircleColors.add(const Color.fromARGB(255, 0, 0, 255));
    mCircleColors.add(const Color.fromARGB(255, 255, 0, 255));
    mCircleColors.add(const Color.fromARGB(255, 255, 0, 0));

    mStatColors.add(const Color.fromARGB(255, 255, 255, 255));
    mStatColors.add(const Color.fromARGB(0, 255, 255, 255));
    hueShader = SweepGradient(colors: mCircleColors);
    saturationShader = RadialGradient(colors: mStatColors);
  }

  @override
  void paint(Canvas canvas, Size size) {
    final rect = Rect.fromLTRB(0.0, 0.0, size.width, size.height);
    mPaint!.shader = hueShader!.createShader(rect);
    saturationPaint!.shader = saturationShader!.createShader(rect);
    // 注意这一句
    canvas.clipRect(rect);
    canvas.drawCircle(
        Offset(size.width / 2, size.height / 2), radius!, mPaint!);
    canvas.drawCircle(
        Offset(size.width / 2, size.height / 2), radius!, saturationPaint!);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return false;
  }
}


lib/circle_page.dart
import 'package:flutter/material.dart';

import 'color_pick.dart';

class CirclePage extends StatefulWidget {
  const CirclePage({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    // TODO: implement createState
    return CirclePageState();
  }
}

class CirclePageState extends State<CirclePage> {
  Color currentColor = const Color(0xff0000ff);

  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: const Text("Color wheel selector"),
      ),
      body: Center(
          child: Column(
        children: <Widget>[
          ColorPickView(
            selectColor: const Color(0xff00eaff),
            selectRadius: 30,
            padding: 100,
            selectRingColor: Colors.black,
            selectColorCallBack: (color) {
              setState(() {
                currentColor = color;
              });
            },
          ),
          Container(
            color: currentColor,
            height: 50,
            width: 50,
            child: const SizedBox(),
          )
        ],
      )),
    );
  }
}


lib/ring_color_pick.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'dart:math';

typedef SelectColor = void Function(Color color);

class ColorRingPickView extends StatefulWidget {
  final Size? size;
  final double? selectRadius;
  final double? padding;
  final double? ringWidth;
  final Color? selectRingColor;
  final Color? selectColor;
  final SelectColor? selectColorCallBack;

  ColorRingPickView(
      {Key? key, this.size,
      this.selectColorCallBack,
      this.selectRadius,
      this.padding,
      this.selectRingColor,
      this.selectColor,
      this.ringWidth}) : super(key: key) {
    assert(size == null || (size != null && size!.height== size!.width),
        'The width and height of the control must be equal');
    assert(size == null || (size != null && size!.height== size!.width),
        'The width and height of the control must be equal');
  }

  @override
  State<StatefulWidget> createState() {
    return ColorPickState();
  }
}

class ColorPickState extends State<ColorRingPickView> {
  double? radius;
  double? selectRadius;
  double? padding;
  double? ringWidth;
  Color currentColor = const Color(0xff00ff);
  Color? selectRingColor;

  Offset? currentOffset;
  Offset? topLeftPosition;
  Offset? selectPosition;
  Size? size;
  Size? screenSize;
  double? centerX, centerY;
  GlobalKey globalKey = GlobalKey();
  bool isTap = false;

  @override
  void initState() {
    super.initState();
    if (widget.selectColor != null && selectPosition == null) {
      Future.delayed(const Duration(milliseconds: 300)).then((val) {
        _setColor(widget.selectColor!);
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    screenSize ??= MediaQuery.of(context).size;
    size = (widget.size ?? screenSize);
    selectRadius = (widget.selectRadius ?? 20);
    padding = (widget.padding ?? 40);
    ringWidth = (widget.ringWidth ?? 20);
    selectRingColor = (widget.selectRingColor ?? Colors.white);
    assert((size != null && screenSize!.width >= size!.width), 'Control width is too wide');
    radius =size!.width / 2 -padding!;
    selectPosition = currentOffset ??= Offset(radius!, radius!);
    _initLeftTop();
    return GestureDetector(
      key: globalKey,
      child: SizedBox(
        width:size!.width,
        height:size!.width,
        child: Stack(
          alignment: Alignment.center,
          children: <Widget>[
            CustomPaint(
              painter: ColorPick(radius: radius, ringWidth:ringWidth),
              size:size!,
            ),
            Positioned(
              left: isTap
                  ? currentOffset!.dx -
                      topLeftPosition!.dx -
                     selectRadius! / 2
                  : selectPosition!.dx -
                      (topLeftPosition == null ? 0 : topLeftPosition!.dx) -
                     selectRadius! / 2,
              top: isTap
                  ? currentOffset!.dy -
                      topLeftPosition!.dy -
                     selectRadius! / 2
                  : selectPosition!.dy -
                      (topLeftPosition == null ? 0 : topLeftPosition!.dy) -
                     selectRadius! / 2,
              child: Container(
                width:selectRadius,
                height:selectRadius,
                decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(widget.selectRadius!),
                    border: Border.fromBorderSide(
                        BorderSide(color:selectRingColor!))),
                child: ClipOval(
                  child: Container(
                    color: currentColor,
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
      onTapDown: (e) {
        setState(() {
          isTap = true;
          _initLeftTop();
          currentColor =
              getColorAtPoint(e.globalPosition.dx, e.globalPosition.dy);
          if (widget.selectColorCallBack != null) {
            widget.selectColorCallBack!(currentColor);
          }
        });
      },
      onPanUpdate: (e) {
        isTap = true;
        _initLeftTop();
        setState(() {
          currentColor =
              getColorAtPoint(e.globalPosition.dx, e.globalPosition.dy);
          print("$centerX-----------$centerY");
          if (widget.selectColorCallBack != null) {
            widget.selectColorCallBack!(currentColor);
          }
        });
      },
    );
  }

  void _initLeftTop() {
    if (globalKey.currentContext != null ) {
      final RenderBox? box =
      globalKey.currentContext!.findRenderObject() as RenderBox?;
      topLeftPosition = box?.localToGlobal(Offset.zero);
      centerX = topLeftPosition!.dx +
         padding! / 2 +
         selectRadius! / 2 +
          radius!;
      centerY = topLeftPosition!.dy +
         padding! / 2 +
         selectRadius! / 2 +
          radius!;
    }
  }

  bool isOutSide(double eventX, double eventY) {
    double x = eventX - centerX!;
    double y = eventY - centerY!;
    double r = sqrt(x * x + y * y);
    if (r >= radius!) return true;
    return false;
  }

  void _setColor(Color color) {
    //设置颜色值
    var hsvColor = HSVColor.fromColor(color);
    _initLeftTop();
    double r = hsvColor.saturation * radius!;
    double radian = hsvColor.hue / 180.0 * pi;
    setState(() {
      currentOffset =
          Offset(centerX! + r * cos(radian), centerY! + r * sin(radian));
      currentColor = color;
    });
  }

  Color getColorAtPoint(double eventX, double eventY) {
    //获取坐标在色盘中的颜色值
    double x = eventX - centerX!;
    double y = eventY - centerY!;
    double r = sqrt(x * x + y * y);
    List<double> hsv = [0.0, 0.0, 1.0];
    var h = (atan2(-y, -x) / pi * 180).toDouble() + 180;
    hsv[0] = h;
    currentOffset =  Offset(centerX! + radius! * cos(h * pi / 180),
        centerY! + radius! * sin(h * pi / 180));
    hsv[1] = max(0, min(1, (r / radius!)));
    return HSVColor.fromAHSV(1.0, hsv[0], hsv[1], hsv[2]).toColor();
  }
}

class ColorPick extends CustomPainter {
  Paint? mPaint;
  Paint? saturationPaint;
  final List<Color> mCircleColors = [];
  final List<Color> mStatColors = [];
  SweepGradient? hueShader;
  final double? radius;
  final double? ringWidth;
  RadialGradient? saturationShader;

  ColorPick({this.radius, this.ringWidth}) {
    _init();
  }

  void _init() {
    //{Color.RED, Color.YELLOW, Color.GREEN, Color.CYAN, Color.BLUE, Color.MAGENTA, Color.RED}
    mPaint =  Paint();
    mPaint?.strokeWidth = ringWidth!;
    mPaint?.style = PaintingStyle.stroke;
    saturationPaint =  Paint();
    mCircleColors.add(const Color.fromARGB(255, 255, 0, 0));
    mCircleColors.add(const Color.fromARGB(255, 255, 255, 0));
    mCircleColors.add(const Color.fromARGB(255, 0, 255, 0));
    mCircleColors.add(const Color.fromARGB(255, 0, 255, 255));
    mCircleColors.add(const Color.fromARGB(255, 0, 0, 255));
    mCircleColors.add(const Color.fromARGB(255, 255, 0, 255));
    mCircleColors.add(const Color.fromARGB(255, 255, 0, 0));

    mStatColors.add(const Color.fromARGB(255, 255, 255, 255));
    mStatColors.add(const Color.fromARGB(0, 255, 255, 255));
    hueShader =  SweepGradient(colors: mCircleColors);
    saturationShader =  RadialGradient(colors: mStatColors);
  }

  @override
  void paint(Canvas canvas, Size size) {
    final rect = Rect.fromLTRB(0.0, 0.0, size.width, size.height);
    mPaint?.shader = hueShader?.createShader(rect);
    saturationPaint?.shader = saturationShader?.createShader(rect);
    // 注意这一句
    canvas.clipRect(rect);
    canvas.drawCircle(Offset(size.width / 2, size.height / 2), radius!, mPaint!);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return false;
  }
}
